### DISCUSSION TITLE
testgrad9.sh to process multiple time steps

### DISCUSSION THREAD

**Discussion by GongCHEN-1995** [TinkerTools/tinker-gpu#164] [Q&A] [answered]
Hi, 

When I use the command "analyze9.sh   test.xyz   test.key   E", if 'test.xyz' contains the positions of one molecule in multiple time steps, it returns molecular energy in different time steps. However, the command "testgrad9.sh test.xyz test.key" only returns the force and energy in the first time step.

Is it possible to make 'testgrad9.sh' be compatible with multiple time steps?

Thanks.

**Comment by zhi-wang** [comment, marked as answer]
It is possible -- there is really no difficulty to achieve this goal. My only concern is that the current output from `testgrad9` is a little different from the Fortran `testgrad` because it does not compute the numerical gradients or the gradients of different energy terms. It is actually a "gradient printer" rather than a comprehensive testing tool as the Fortran `testgrad`. Adding this feature will make them even more different.

A workaround using a Bash script is easy, if you also use the Fortran Tinker. There is an option in executable `archive`, option 2 if my recollection is correct, to extract individual frames from an archive file. This bash file will look similar to this example and some modification may also be necessary:

```
archive molecules.arc 2 [more args]
for x in molecules.[0-9][0-9][0-9]
do
testgrad9.sh $x
done
```

**Reply by GongCHEN-1995** [reply]
Thanks. I have tried your advice. "archive" creats a file for every time step. So my problem is solved.

But I am still curious about how "analyze9.sh" processes the multiple time steps. Does it also divide the file as your example and process them one by one? Or "analyze9.sh" transfers data (of all time setps) to GPU and process them simultaneously?

**Reply by zhi-wang** [reply]
The gist of `analyze` is the same as the previous example, but the implementation does not involve calling `archive` or interacting with shell commands. The trajectory file is parsed by the C++ code and the energy evaluation is called between the frames.

**Reply by GongCHEN-1995** [reply]
I get it. Thanks

**Reply by GongCHEN-1995** [reply]
I find that it is too slow if I use 'testgrad9.sh' to process the time steps one by one. So I make a little change to the file "tinker9/src/testgrad_x.cpp" in your source code. Now 'testgrad9.sh' is able to return the results of multiple time steps in one time. It is faster than processing one time step at a time. Here is the code for "tinker9/src/testgrad_x.cpp". 


#include "energy.h"
#include "tinker_rt.h"
#include "tool/io_text.h"
#include <tinker/detail/inform.hh>
#include <fstream>
#include <tinker/detail/files.hh>
#include "nblist.h"


#define TINKER_TESTGRAD_VIRIAL 0


namespace tinker {
void x_testgrad_one();

void x_testgrad(int, char**)
{
   initial();
   TINKER_RT(getxyz)();
   mechanic();
   mechanic2();


   auto out = stdout;


   int flags = (calc::xyz + calc::mass);
   flags += (calc::energy + calc::grad);
#if TINKER_TESTGRAD_VIRIAL
   flags += calc::virial;
#endif


   rc_flag = flags;
   initialize();
   energy(rc_flag);

   fstr_view fsw = files::filename;
   std::string fname = fsw.trim();
   std::ifstream ipt(fname);
   int done = false;
   read_frame_copyin_to_xyz(ipt, done);
   refresh_neighbors();
   int nframe_processed = 1;
   x_testgrad_one();

   while (!done) {
      read_frame_copyin_to_xyz(ipt, done);
      refresh_neighbors();
      nframe_processed++;
      print(out, "\n Analysis for Archive Structure :%16d\n", nframe_processed);
      x_testgrad_one();
   }
   finish();
   TINKER_RT(final)();

}

void x_testgrad_one()
{
    energy(calc::energy + calc::grad);
    auto& out = stdout;
    int digits = inform::digits;
    std::vector<double> gdx(n), gdy(n), gdz(n);
    copy_gradient(calc::grad, gdx.data(), gdy.data(), gdz.data());
    
    const int len_e = 20 + digits;
    const char* fmt_e = "\n Total Potential Energy :%1$*2$.*3$f Kcal/mole\n\n";
    print(out, fmt_e, esum, len_e, digits);

 #if TINKER_TESTGRAD_VIRIAL
    const char* fmt_v = " %-36s%12.3f %12.3f %12.3f\n";
    print(out, fmt_v, "Internal Virial Tensor :", vir[0], vir[1], vir[2]);
    print(out, fmt_v, "", vir[3], vir[4], vir[5]);
    print(out, fmt_v, "", vir[6], vir[7], vir[8]);
 #endif

    print(out, "Cartesian Gradient Breakdown over Individual Atoms :");
    
    std::string fmt_t;
    std::string fmt;
    if (digits == 8) {
       fmt_t = "\n  Type    Atom %1$8s "
               "dE/dX %1$9s dE/dY %1$9s dE/dZ %1$9s Norm\n";
       fmt = "\n Anlyt%8d %16.8f %16.8f %16.8f %16.8f";
    } else if (digits == 6) {
       fmt_t = "\n  Type      Atom %1$9s "
               "dE/dX %1$7s dE/dY %1$7s dE/dZ %1$9s Norm\n";
       fmt = "\n Anlyt%10d   %14.6f %14.6f %14.6f %14.6f";
    } else {
       fmt_t = "\n  Type      Atom %1$12s "
               "dE/dX %1$5s dE/dY %1$5s dE/dZ %1$8s Norm\n";
       fmt = "\n Anlyt%10d       %12.4f %12.4f %12.4f %12.4f";
    }
    print(out, fmt_t, "");
    // auto do_print = [](int i, int n, int top_m) {
    //    if (n <= 2 * top_m)
    //       return true;
    //    else if (i < top_m)
    //       return true;
    //    else if (i >= n - top_m)
    //       return true;
    //    else
    //       return false;
    // };
    // int print_top_n = 15;
    for (int i = 0; i < n; ++i) {
       // if (not do_print(i, n, print_top_n))
       //    continue;


       real x1 = gdx[i];
       real y1 = gdy[i];
       real z1 = gdz[i];
       real norm = x1 * x1 + y1 * y1 + z1 * z1;
       norm = REAL_SQRT(norm);
       print(out, fmt, i + 1, x1, y1, z1, norm);
    }


    double totnorm = 0;
    for (int i = 0; i < n; ++i) {
       totnorm += gdx[i] * gdx[i] + gdy[i] * gdy[i] + gdz[i] * gdz[i];
    }
    totnorm = std::sqrt(totnorm);
    double rms = totnorm / std::sqrt(n);
    print(out, "\n\n Total Gradient Norm and RMS Gradient per Atom :\n");
    const char* fmt3 = "\n Anlyt      %1$-30s%2$*3$.*4$f\n";
    const int len3 = 13 + digits;
    print(out, fmt3, "Total Gradient Norm Value", totnorm, len3, digits);
    print(out, fmt3, "RMS Gradient over All Atoms", rms, len3, digits);
}
}

**Reply by zhi-wang** [reply]
You can send us a pull request if you are interested.
